### PR TITLE
Selenium test

### PR DESCRIPTION
**Pull request by svenjens**
Description 
Added e2e test for ie, chrome, firefox, safari with nightwatch on browserstack. 
Please make sure:
- [x] to add a label
- [x] to add one or more reviewers
- [x] you followed our checklist for [functional testing](https://github.com/Amsterdam/vergunningcheck/blob/master/TESTING.md)
- [x] you added the necessary documentation (either in the markdown files or inline)
- [x] you added the necessary automated tests

### CODE REVIEW

File: packages/e2e/nightwatch/tests/imtrFlow.js [hunk 1]
```diff
@@ -37,14 +37,17 @@ module.exports = {
       "Voer het adres in waar u de dakkapel wilt gaan plaatsen"
     );
     assert.visible(locationPostalCode);
-    b.setValue(locationPostalCode, address.zipCode);
+    b.setValue(locationPostalCode, '1055XD');
+    // b.setValue(locationPostalCode, address.zipCode);
```
Comment (afjlambert) [inline packages/e2e/nightwatch/tests/imtrFlow.js]: Why did you change this?

File: packages/e2e/nightwatch/browserstack.conf.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
+const browserstack = require('browserstack-local');
```
Comment (afjlambert) [inline packages/e2e/nightwatch/browserstack.conf.js]: Not needed?

File: packages/e2e/address.test.js [hunk 3]
```diff
@@ -1,18 +1,18 @@
-describe("mocking support", () => {
```
Comment (afjlambert) [inline packages/e2e/address.test.js:1]: Why did you comment this?
Please note the `it` is x'ed (`xid` so the test is skipped anyway)

File: package.json [hunk 4]
```diff
@@ -39,7 +39,8 @@
   },
   "husky": {
     "hooks": {
-      "pre-commit": "lerna run --concurrency 1 --stream precommit"
+      "pre-commit": "lerna run --concurrency 1 --stream precommit",
+      "pre-push": "lerna run --concurrency 1 --stream nightwatch:browserstack"
```
Comment (afjlambert) [inline package.json]: Nice. We need something similar on the build server.

File: packages/e2e/address.js [hunk 5]
```diff
@@ -1,3 +1,3 @@
-const { address1055 } = require("@vergunningcheck/mocking/src/address");
```
Comment (afjlambert) [inline packages/e2e/address.js:1]: why?

File: packages/e2e/package.json [hunk 6]
```diff
@@ -31,5 +32,10 @@
     "selenium-server": "^3.141.59",
     "ts-node": "^9.0.0",
     "typescript": "^4.0.3"
+  },
+  "dependencies": {
```
Comment (afjlambert) [inline packages/e2e/package.json:38]: Did you add them to dependencies deliberately? Consider moving them to devDependencies.

File: packages/e2e/scripts/local.runner.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+          // Code to stop browserstack local after end of parallel test
+          bs_local.stop(() => {});
+        })
+        .runTests(() => {
+          // Code to stop browserstack local after end of single test
+          bs_local.stop(() => {});
+        });
+    });
+  });
+} catch (ex) {
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js]: I think we use `e` as the exception/error-variable name in most places.

File: packages/e2e/scripts/local.runner.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env node
+const Nightwatch = require('nightwatch');
+const browserstack = require('browserstack-local');
+let bs_local;
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js]: Can be moved to the try block as it is not used in the catch anyway.

File: packages/e2e/scripts/local.runner.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        .runTests(() => {
+          // Code to stop browserstack local after end of single test
+          bs_local.stop(() => {});
+        });
+    });
+  });
+} catch (ex) {
+  console.log('There was an error while starting the test runner:\n\n');
+  process.stderr.write(ex.stack + '\n');
+  process.exit(2);
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js:26]: Why exit 2?

File: packages/e2e/scripts/local.runner.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+const browserstack = require('browserstack-local');
+let bs_local;
+
+try {
+
+  process.mainModule.filename = "./node_modules/nightwatch/bin/nightwatch"
+  // Code to start browserstack local before start of test
+  console.log("Connecting local");
+  Nightwatch.bs_local = bs_local = new browserstack.Local();
+  bs_local.start({'key': process.env.BROWSERSTACK_ACCESS_KEY }, (error) => {
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js:13]: What if you don't have this variable in your env? Or is it guaranteed?

File: packages/e2e/scripts/local.runner.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+const Nightwatch = require('nightwatch');
+const browserstack = require('browserstack-local');
+let bs_local;
+
+try {
+
+  process.mainModule.filename = "./node_modules/nightwatch/bin/nightwatch"
+  // Code to start browserstack local before start of test
+  console.log("Connecting local");
+  Nightwatch.bs_local = bs_local = new browserstack.Local();
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js:12]: Is all this needed?

File: packages/e2e/scripts/local.runner.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  bs_local.start({'key': process.env.BROWSERSTACK_ACCESS_KEY }, (error) => {
+    if (error) throw error;
+
+    console.log('Connected. Now testing...');
+    Nightwatch.cli(function(argv) {
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js]: Can we use error functions here?

File: packages/e2e/scripts/local.runner.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+      Nightwatch.CliRunner(argv)
+        .setup(null, () => {
+          // Code to stop browserstack local after end of parallel test
+          bs_local.stop(() => {});
+        })
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js]: Same anonymous function is used twice. Consider extracting it as a named function and `.setup(null, someFunc)`

File: packages/e2e/scripts/local.runner.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        .runTests(() => {
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js]: runTests(someFunctionName)...

File: packages/e2e/scripts/local.runner.js [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+          // Code to stop browserstack local after end of single test
+          bs_local.stop(() => {});
+        });
+    });
+  });
+} catch (ex) {
+  console.log('There was an error while starting the test runner:\n\n');
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js]: I'd use console.error

File: spikes/new-form-view.md [hunk 16]
```diff
@@ -0,0 +1,60 @@
+## Alles onder elkaar ipv achter elkaar
```
Comment (afjlambert) [inline spikes/new-form-view.md]: Is this part of the selenium tests?

File: spikes/sttr-forms-without-outcome.md [hunk 17]
```diff
@@ -0,0 +1,10 @@
+### STTR Files zonder outcome
```
Comment (afjlambert) [inline spikes/sttr-forms-without-outcome.md]: Is this part of the selenium tests?

**Review by afjlambert — CHANGES_REQUESTED** [review]
Looks great @svenjens 
Some comments to see if we can improve this further.

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -107,7 +107,7 @@ If you want to combine 2 or more permits in 1 checker, we need to configure a to
 
 - Follow the install steps in [./packages/imtr/README.md](./scripts/sttimtrr_build/README.md)
 - First make sure you update your permits so we have all the files we need. See the section above.
-- List the permit-id's you want to combine in [./packages/imtr/src/config.ts](./packages/imtr/src/config.ts) under `topics`.
+- List the permit-id's you want to combine in [./packages/imtr/src/config.js](./packages/imtr/src/config.ts) under `topics`.
```
Comment (afjlambert) [inline CONTRIBUTING.md]: This is not right I think.

File: Jenkinsfile [hunk 19]
```diff
@@ -33,6 +35,17 @@ node {
         checkout scm
     }
 
+    if (BRANCH == "docker-test") {
```
Comment (afjlambert) [inline Jenkinsfile]: Can we move this to the section after deployment?

File: Jenkinsfile [hunk 20]
```diff
@@ -33,6 +35,17 @@ node {
...
+        stage('Run e2e test') {
+            tryStep "Run e2e test", {
+                docker.withRegistry("${DOCKER_REGISTRY_HOST}",'docker_registry_auth') {
+                    e2e_image = docker.build("${NIGHTWATCH_CONTAINERNAME}","--pull --build-arg=JENKINS_URL=${env.JENKINS_URL} --build-arg=JOB_NAME=${env.JOB_NAME} --build-arg=BUILD_URL=${env.BUILD_URL} --build-arg=BUILD_NUMBER=${env.BUILD_NUMBER} -f ${NIGHTWATCH_DOCKERFILE} ${CONTAINERDIR}")
+    //              e2e_image.push()
```
Comment (afjlambert) [inline Jenkinsfile]: Please remove comment if not needed.

File: ci/Dockerfile.nightwatch [hunk 21]
```diff
@@ -0,0 +1,11 @@
+FROM node:alpine
+WORKDIR /app/
+
+# Install
+COPY . ./
+
+RUN rm -rf packages/client packages/graphql packages/imtr-client
```
Comment (afjlambert) [inline ci/Dockerfile.nightwatch]: Can you add a comment why we do this?
There were some conflicts in the packages? Can't remember.

File: package.json (hunk 4 above)
Comment (afjlambert) [inline package.json]: This command runs nightwatch:browserstack agains production. Why would we do this pre-push?

File: packages/e2e/address.js [hunk 22]
```diff
@@ -1,3 +1,9 @@
...
-
+// const { address1055 } = require("@vergunningcheck/mocking/src/address");
```
Comment (afjlambert) [inline packages/e2e/address.js]: Can you replace this comment? Maybe explain why we don't import from the mocking package here. Therefore some duplicate code here is acceptable.

File: packages/e2e/nightwatch/browserstack.conf.js (hunk 2 above)
Comment (afjlambert) [reply packages/e2e/nightwatch/browserstack.conf.js]: It seems this variable is not used here. Is it an unneeded require or does it setup some global stuff? Please try without this require. If it fails then maybe add a comment and eslint-ignore it.

File: packages/e2e/nightwatch/browserstack.conf.js [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+  const config = nightwatch_config.test_settings[i];
+  config['selenium_host'] = nightwatch_config.selenium.host;
+  config['selenium_port'] = nightwatch_config.selenium.port;
+  config['desiredCapabilities'] = config['desiredCapabilities'] || {};
+  for(let j in nightwatch_config.common_capabilities){
+    config['desiredCapabilities'][j] = config['desiredCapabilities'][j] || nightwatch_config.common_capabilities[j];
+  }
+}
+
+module.exports = nightwatch_config;
```
Comment (afjlambert) [inline packages/e2e/nightwatch/browserstack.conf.js]: Please add a newline.

File: packages/e2e/package.json (hunk 6 above)
Comment (afjlambert) [reply packages/e2e/package.json:38]: Still stands.

File: packages/e2e/scripts/local.runner.js (hunk 8 above)
Comment (afjlambert) [reply packages/e2e/scripts/local.runner.js]: Same here.

File: packages/e2e/nightwatch/tests/imtrFlow.js [hunk 24]
```diff
@@ -1,9 +1,7 @@
 const { address, domain, selectors } = require("../config");
```
Comment (afjlambert) [inline packages/e2e/nightwatch/tests/imtrFlow.js]: I see a lot of changes in this file. Can you revert them?

**Review by afjlambert — CHANGES_REQUESTED** [review]

File: packages/e2e/scripts/local.runner.js [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+const browserstack = require('browserstack-local');
+let bs_local;
+
+try {
+
+  process.mainModule.filename = "./node_modules/nightwatch/bin/nightwatch"
+  // Code to start browserstack local before start of test
+  console.log("Connecting local");
+  Nightwatch.bs_local = bs_local = new browserstack.Local();
+  bs_local.start({'key': '3T6jZ9N4pYf9qjzoqKxU' }, (error) => {
```
Comment (afjlambert) [inline packages/e2e/scripts/local.runner.js]: Please remove this.

**Review by afjlambert — CHANGES_REQUESTED** [review]

File: packages/e2e/package.json [hunk 26]
```diff
@@ -6,16 +6,15 @@
   "keywords": [],
   "author": "Gemeente Amsterdam",
   "license": "ISC",
-  "main": "puppeteer/basic.ts",
-  "type": "module",
+  "main": "puppeteer/basic.js",
   "scripts": {
     "test": "jest",
-    "nightwatch:browserstack": "node --experimental-modules scripts/local.runner.js -c nightwatch/browserstack.conf.js -e chrome,firefox,safari,ie",
+    "nightwatch:browserstack": "node scripts/local.runner.js -c nightwatch/browserstack.conf.js -e chrome",
```
Comment (afjlambert) [inline packages/e2e/package.json]: can you re-add the `firefox,safari,ie` ?

File: packages/e2e/scripts/local.runner.js (hunk 10 above)
Comment (svenjens) [reply packages/e2e/scripts/local.runner.js:13]: i will update the docs

Comment (afjlambert) [reply packages/e2e/scripts/local.runner.js:13]: Still open?

File: Jenkinsfile [hunk 27]
```diff
@@ -9,6 +9,8 @@ String PLAYBOOK = 'deploy.yml'
 // All other data uses variables, no changes needed for static
 String NGINX_CONTAINERNAME = "ois/vergunningcheck_nginx:${env.BUILD_NUMBER}"
 String NGINX_DOCKERFILE="ci/Dockerfile.client"
+String NIGHTWATCH_CONTAINERNAME = "ois/vergunningcheck_nightwatch:${env.BUILD_NUMBER}"
+String NIGHTWATCH_DOCKERFILE="ci/Dockerfile.nightwatch"
```
Comment (afjlambert) [inline Jenkinsfile]: Can these be removed?

File: ci/Dockerfile.nightwatch [hunk 28]
```diff
@@ -0,0 +1,11 @@
+FROM node:alpine
```
Comment (afjlambert) [inline ci/Dockerfile.nightwatch]: This file can be removed?

File: packages/e2e/.gitignore [hunk 29]
```diff
@@ -1,3 +1,4 @@
 build/
 tests_output/
+.env
```
Comment (afjlambert) [inline packages/e2e/.gitignore]: Please remove. The root of our project is not a python project. You probably had this issue locally. `.env` should only be excluded in the manager app.

File: CONTRIBUTING.md [hunk 30]
```diff
@@ -168,7 +168,10 @@ When we want to test our app with users we most follow this procedure:
 Basically what we want to do is merge `develop` with `release` including the latest IMTR-changes. We use lerna-changelog to generate our changes we can use in [CHANGELOG.md](CHANGELOG.md), so you'll need a [personal access token](https://github.com/settings/tokens) for the GitHub API with the public_repo scope for public repositories.
 Make sure you are logged in by npm command line. If not, log in with `npm adduser`. Add `export GITHUB_AUTH=...` to your profile (eg: `.zshrc`).
 
-- Run `npm run prepare-release` to back-merge and generate new IMTR-files
+- Run the project locally on http://localhost:3000/. 
+- In new tab run `npm run e2e-browserstack` to test if the application is running in all browsers.
+- Close the running application.
+- In new tab run `npm run prepare-release` to back-merge and generate new IMTR-files
```
Comment (afjlambert) [inline CONTRIBUTING.md:174]: Can you move this to after deployment to acceptance and test it on that environment as well? You are now testing your local dev-env and not the production build. Thanks

File: packages/e2e/nightwatch/browserstack.conf.js [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+
+  selenium : {
+    "start_process" : false,
+    "host" : "hub-cloud.browserstack.com",
+    "port" : 80
+  },
+
+  common_capabilities: {
+    'build': 'nightwatch-browserstack',
+    'browserstack.user': 'sjens1',
```
Comment (afjlambert) [inline packages/e2e/nightwatch/browserstack.conf.js]: Can you make this an env-var as well? And document how to obtain it?

File: packages/e2e/nightwatch/tests/imtrFlow.js (hunk 24 above)
Comment (afjlambert) [reply packages/e2e/nightwatch/tests/imtrFlow.js]: Please undo the changes or explain why you changed it.
Also there is a newline missing on the last line. Did you setup prettier/editorconfig correctly?

File: packages/e2e/nightwatch/tests/oloFlow.js [hunk 32]
```diff
@@ -20,19 +21,14 @@ module.exports = {
...
     assert.visible(locationHouseNumberFull);
     b.setValue(locationHouseNumberFull, address.houseNumberFull);
     b.waitForElementVisible(locationFound);
     assert.containsText(main, address.streetName);
 
-    // forward, back, forward
-    b.click(navButtonNext);
-    b.waitForElementVisible(navButtonPrev);
-    b.click(navButtonPrev);
-    b.waitForElementVisible(navButtonNext);
```
Comment (afjlambert) [inline packages/e2e/nightwatch/tests/oloFlow.js]: Why did you remove this?

File: packages/e2e/package.json [hunk 33]
```diff
@@ -23,13 +24,20 @@
     "chai": "^4.2.0",
     "chromedriver": "^87.0.5",
     "geckodriver": "^1.20.0",
-    "jest": "^26.6.3",
+    "jest": "^26.5.3",
```
Comment (afjlambert) [inline packages/e2e/package.json]: Did you downgrade jest on purpose?

File: packages/e2e/package.json [hunk 34]
```diff
@@ -23,13 +24,20 @@
...
+    "prettier": "^2.1.2",
+    "ts-jest": "^26.4.1",
```
Comment (afjlambert) [inline packages/e2e/package.json]: I think you don't need ts-jest anymore.

File: packages/e2e/tsconfig.json [hunk 35]
```diff
@@ -4,9 +4,12 @@
```
Comment (afjlambert) [inline packages/e2e/tsconfig.json]: Please remove this entire file.

File: packages/e2e/scripts/local.runner.js (hunk 15 above)
Comment (afjlambert) [reply packages/e2e/scripts/local.runner.js]: You don't agree?

File: packages/e2e/scripts/local.runner.js (hunk 13 above)
Comment (afjlambert) [reply packages/e2e/scripts/local.runner.js]: You don't agree?

**Review by afjlambert — CHANGES_REQUESTED** [review]

File: README.md [hunk 36]
```diff
@@ -13,6 +13,15 @@ npm start
 
 If you need to update the IMTR files there are additional install steps in [packages/imtr/README.md](packages/imtr/README.md).
 
+If you want to use browserstack cloud service:
+you will need the following configuration 
+Please set the username and access key by setting the environment variables: 
+  - BROWSERSTACK_USER                                                          
+  - BROWSERSTACK_KEY
```
Comment (afjlambert) [inline README.md]: Please check `BROWSERSTACK_USER` vs. `BROWSERSTACK_USERNAME`.

Please check `BROWSERSTACK_KEY` vs. `BROWSERSTACK_ACCESS_KEY`

File: packages/e2e/nightwatch/tests/oloFlow.js (hunk 32 above)
Comment (afjlambert) [reply packages/e2e/nightwatch/tests/oloFlow.js]: Is it possible to add them again? Seems like a good test.

**Review by afjlambert — CHANGES_REQUESTED** [review]
